### ISSUE TITLE
instantiation+implicit conversion for numeric types

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#20011]
Today, the compiler supports certain patterns of instantiation + implicit conversion for class types. However it does not support it for numeric types.

Here is an example that would need implicit conversions + instantiation with numeric types in order to compile:

``` chapel
proc f(x: chpl_anyreal, y: chpl_anyreal) {
  writeln("  f(", x.type:string, ",", y.type:string, ")");
}

var myInt: int = 1;
var myUint: uint = 2;
f(myInt, myInt);
f(myUint, myUint);
f(myInt, myUint);
```

This does not compile since the first `f` call fails to resolve. However, the compiler could know that, since `int` can implicitly convert to `real`, it can instantiate `f` with `real` and pass to that instantiation.

We have wanted to replace the stamping-out for each integer width currently done with e.g. `proc g(arg: int(?w))` with some sort of generic argument processing. Doing so in the most obvious would run into this issue (and potentially other issues, as well).

See also https://github.com/chapel-lang/chapel/issues/5054#issuecomment-282748366 which discusses an example where this kind of functionality would be relevant if/when we add user-defined implicit conversions.

**Associated Future Test(s):**
[`test/types/coerce/allNumericsBinaryGeneric.chpl`](
           https://github.com/chapel-lang/chapel/blob/main/test/types/coerce/allNumericsBinaryGeneric.chpl
          ) #19959 (if this issue is fixed the generic overload would be called in fewer cases in this test)

**Comment by bradcray** [comment]
> We have wanted to replace the stamping-out for each integer width currently done with e.g. proc g(arg: int(?w)) with some sort of generic argument processing. Doing so in the most obvious would run into this issue (and potentially other issues, as well).

I'm not understanding the relationship to this issue.  Can you clarify?

To state my take on the `proc g(arg: int(?w))` issue , I believe that, today, when we see such routines, we proactively stamp them out for each of the integer widths rather than instantiating on demand based on callsites, which has always seemed odd to me.  Specifically, if `int` were just a normal user record, we wouldn't instantiate it for all possible generic fields (because there'd be an arbitrary number of such signatures), so it seems asymmetrical.  It also seems like it potentially creates more overloads than a given program may want or need, which seems unfortunate (though I suppose it may also "win" in compilation time overall since by auto-expanding, we're resolving against 4 concrete overloads rather than dealing with a generic?)

**Comment by mppf** [comment]
> > We have wanted to replace the stamping-out for each integer width currently done with e.g. proc g(arg: int(?w)) with some sort of generic argument processing. Doing so in the most obvious would run into this issue (and potentially other issues, as well).
> 
> I'm not understanding the relationship to this issue. Can you clarify?

Well, if we wrote this

``` chapel
proc f(x: real(?w), y: real(w)) {
  writeln("  f(", x.type:string, ",", y.type:string, ")");
}

var myInt: int = 1;
var myUint: uint = 2;
f(myInt, myInt);
```

it would compile just fine. But the way I was imagining it would be "obvious" to convert the `real(?w)` into a generic function creates this:

``` chapel
proc f(x: chpl_anyreal, y: chpl_anyreal) {
  writeln("  f(", x.type:string, ",", y.type:string, ")");
}

var myInt: int = 1;
var myUint: uint = 2;
f(myInt, myInt);
```

which does not compile for the reasons described in the O.P.

Now, you might have expected it to be this:
``` chapel
proc f(x: chpl_anyreal, y: x.type) {
  writeln("  f(", x.type:string, ",", y.type:string, ")");
}
```

however that would presumably cause the function to only compute the real type to instantiate from the 1st argument rather than considering both. E.g. `f(myReal32, myReal64)` would try to call the function instantiated with `real(32)` but in fact we want it to be instantiated with `real(64)`. That is why I used the signature `proc f(x: chpl_anyreal, y: chpl_anyreal) ` in the original post but of course we might decide upon a different way of solving this problem.



> To state my take on the `proc g(arg: int(?w))` issue , I believe that, today, when we see such routines, we proactively stamp them out for each of the integer widths rather than instantiating on demand based on callsites, which has always seemed odd to me. Specifically, if `int` were just a normal user record, we wouldn't instantiate it for all possible generic fields (because there'd be an arbitrary number of such signatures), so it seems asymmetrical. It also seems like it potentially creates more overloads than a given program may want or need, which seems unfortunate (though I suppose it may also "win" in compilation time overall since by auto-expanding, we're resolving against 4 concrete overloads rather than dealing with a generic?)

An in-between strategy would be to create the 4 concrete overloads on-demand (at the same time we would normally instantiate) instead of proactively. That would meet some of the desires you mentioned without running into the problems of instantiation+implicit conversion or argument order.

**Comment by mppf** [comment]
> 
> Well, if we wrote this
> 
> ```chapel
> proc f(x: real(?w), y: real(w)) { ... }
> ```
> 
> it would compile just fine. But the way I was imagining it would be "obvious" to convert the `real(?w)` into a generic function creates this:
> 
> ```chapel
> proc f(x: chpl_anyreal, y: chpl_anyreal) { ... }
> ```

> An in-between strategy would be to create the 4 concrete overloads on-demand (at the same time we would normally instantiate) instead of proactively. That would meet some of the desires you mentioned without running into the problems of instantiation+implicit conversion or argument order.

Maybe we just need special logic in the resolver to handle this case? When `?t` is "simple enough", it can consider `proc f(x: real(?w), y: real(w))` as meaning "both arguments are real of the same width" and so given actuals with two different real widths, maybe it could instantiate it both ways? Then it is just stamping out 2 functions based on the call site instead of more (well there are only 2 real widths today, but there would be 4 for an integer type).

> 
> which does not compile for the reasons described in the O.P.
> 
> Now, you might have expected it to be this:
> 
> ```chapel
> proc f(x: chpl_anyreal, y: x.type) {
>   writeln("  f(", x.type:string, ",", y.type:string, ")");
> }
> ```
> 
> however that would presumably cause the function to only compute the real type to instantiate from the 1st argument rather than considering both. E.g. `f(myReal32, myReal64)` would try to call the function instantiated with `real(32)` but in fact we want it to be instantiated with `real(64)`. That is why I used the signature `proc f(x: chpl_anyreal, y: chpl_anyreal) ` in the original post but of course we might decide upon a different way of solving this problem.
> 
> > To state my take on the `proc g(arg: int(?w))` issue , I believe that, today, when we see such routines, we proactively stamp them out for each of the integer widths rather than instantiating on demand based on callsites, which has always seemed odd to me. Specifically, if `int` were just a normal user record, we wouldn't instantiate it for all possible generic fields (because there'd be an arbitrary number of such signatures), so it seems asymmetrical. It also seems like it potentially creates more overloads than a given program may want or need, which seems unfortunate (though I suppose it may also "win" in compilation time overall since by auto-expanding, we're resolving against 4 concrete overloads rather than dealing with a generic?)
> 
> An in-between strategy would be to create the 4 concrete overloads on-demand (at the same time we would normally instantiate) instead of proactively. That would meet some of the desires you mentioned without running into the problems of instantiation+implicit conversion or argument order.

**Comment by bradcray** [comment]
After some conversations this week, my current thought on the current "have `real(?w)` formals result in stamping out the routine for all legal values of `w`" behavior is that we should disable this capability, most likely by changing declarations like:

```chapel
operator +(x: real(?w), y: real(w)) { ... }
```

into the explicit series of instantiations, say:

```chapel
operator +(x: real(32), y: real(32)) { ... }
operator +(x: real(64), y: real(64)) { ... }
```

And then, more generally, when we see `proc foo(x: real(?w), y: real(w))` we should treat it the way it appears:  The first argument is a generic that is expected to be a real, and whatever its width is, that constrains the second argument as well.  Thus, if I wrote:  `foo(myReal32, myReal64)`, `w` would be 32 and the call would not resolve.  But if I wrote `foo(myReal64, myReal32)` it would work fine (promoting `myReal32` to a `real(64)`, as would `foo(myReal32, myInt64)`.

My current thinking is that we should not permit an integer argument to resolve to a `real(?w)` formal because the actual is not a real and the routine may be doing some sort of float-specific operation on it.  For example, imagining:

```chapel
proc numBitsInExponent(x: real(?w)) { ... }
```

it doesn't really make sense to call `numBitsInExponent(3)` even though 3 could coerce to 3.0 and an answer could be produced for the resulting real temporary.  If the author of `numBitsInExponent()` wanted it to handle integer arguments as well, they should either make it non-generic or add integer overloads.

At least, that's my current thought in terms of making the language less surprising and to make the `: real(?w)`-style arguments less special/magical.  I'd at least be interested in seeing what the impact on the code base would be.

**Comment by mppf** [comment]
I agree that trying that out & seeing the impact on the code base is a good next step.

I am on the uncertain about what would be more intiutive. Should we be trying to make


```
proc f(arg: int(?w))
```

and

```
proc f(arg: int(8))
proc f(arg: int(16))
proc f(arg: int(32))
proc f(arg: int(64))
```

behave the same in terms of what can be passed to them? Having them behave the same seems intuitive to me (but perhaps that is just because I am used to it with the current behavior).

**Comment by bradcray** [comment]
> behave the same in terms of what can be passed to them?

I'm currently thinking "no".  `int(?w)` seems to say "I can take any integer" and `bool` is not an integer.

**Comment by mppf** [comment]
(In the below I am assuming that we make `int(?w)` a real generic type; so it would be equivalent to a `chpl_anyInt` type).

I'm thinking that what we should aim for here is that it be possible to write both patterns:
 1. When it is desired that implicit conversions can be applied while also instantiating
 2. When it is desired that no such implicit conversions occur.

Anyway, how could we express both of these behaviors depending on our choice here?

### If `int(?w)` does allow implicit conversion

``` chapel
// Case 1
proc acceptsAnythingConvertableToInt(arg: int(?w)) { writeln(arg); }
// Case 2
proc acceptsOnlyInts(arg) where isIntType(arg.type) { writeln(arg); }
```

Here I am expecting that, when the compiler is choosing which `?w` to instantiate with for `acceptsAnythingConvertableToInt`, it would do so in a manner consistent with choosing among overloads with all of the `int` widths. For example, if we pass `true`, since `bool` prefers to pass to a default-sized `int`, it would choose to instantiate `?w` with 64 to get a default sized int. This might even depend on a param's value (although I don't have a concrete case in mind that would do that in practice).

### If `int(?w)` does not allow implicit conversion

``` chapel
// Case 1
proc acceptsAnythingConvertableToInt(arg) where isCoercible(arg.type, int) {
  var x = arg : computePreferredIntType(arg);
}

// Case 2
proc acceptsOnlyInts(arg: int(?w)) {
  writeln(arg);
}
```

This approach has some downsides for Case 1:
 1. It's not totally clear what destination type to put in the where clause for `isCoercible`. `int` doesn't seem quite right but might work in practice.
 2. `computePreferredIntType` would presumably need to be added to the standard library
 3. The code for Case 1 is actually fairly complex


Another concern is that some of the implicit conversions depend on information from the call site that would be lost inside of the body of a function. In particular, we have the ability for `1` which has type `int` to implicitly convert to `int(32)`, that is not something that can apply within the body of the function unless it accepts the argument by `param`. However, I am having trouble coming up with practical examples that suffer from this problem.

### Another Direction

A totally different direction is to allow both behaviors by having different generic numeric types available in the standard library. E.g. `anyIntAllowingConversions` could be a type in addition to `int(?w)` aka `chpl_anyInt`.

**Comment by mppf** [comment]
(This is a somewhat related aside, AFAIK it doesn't impact the language design discussion).


In some development on dyno we noticed the following test case, which doesn't use the standard library at all, but causes the dyno resolver to encounter a recursive query:

``` chapel
module M {
  operator *(a: real(?w), b: complex(w*2)) { return 0; }
  operator *(a: int(?w), b: int(w)) { return 0; }
  proc main() {
    var x = 1*2;
  }
}
```

It is odd that resolving `1*2` seems to cause a loop:
 * finding and instantiating candidates leads it to try to resolve the `real`/`complex` version (which is a candidate with implicit conversions)
 * doing so causes it to try to resolve another `*` for the type expression  `complex(w*2)`
 * at that point the dyno resolver fails because it cannot handle recursive queries.

**Comment by vasslitvinov** [comment]
The changes in #22348 make our code even more bloated, including clones of the cases like:
```chpl
  inline operator <(a: uint(8), param b: uint(8)) {
    if __primitive("==", b, 0) {
      return false;
    } else {
      return __primitive("<", a, b);
    }
  }

  iter chpl_direct_strided_range_iter(low: uint(8), high: uint(8),
                                      stride: integral) {
    const r = low..high by stride;
    for i in r do yield i;
  }
```

I am game saying "this does not affect our users, we will improve it as time allows." Yet, it makes me want to have built-in support for binary operators that use C++ rules for situations like int8 + uint32, as well as for formals and locals that can be either `param` or `const` depending on what is passed to them.

**Comment by mppf** [comment]
I just came across this proposal from the [constrained generics proposal CHIP 2](https://github.com/chapel-lang/chapel/blob/main/doc/rst/developer/chips/2.rst) which seems to argue for allowing the combination of instantiation and implicit conversion.

> ### Query Types
> 
> We propose relaxing the restriction that a query type may only appear once in a
> function signature. Instead, a query type may appear any number of times. The
> query type resolves to the match for the first occurrence of the query type.
> 
> #### Example
> 
> 
> The following function and function call are well typed, assuming that
> real numbers implement LessThanComparable. The query type T resolves to real,
> and the second argument in the function call is coerced to real.
> 
> ``` chapel
>   proc min(x : ?T, y : ?T): T where T implements LessThanComparable {
>     if y < x then
>       return y;
>     else
>       return x;
>   }
>   min(1.0, 2);
> ```

**Comment by mppf** [comment]
This comment summarizes information this issue to prepare for making a decision

### Summary of the question

We are trying to decide if a generic function can also rely on instantiation.
Here is an example:
``` chapel
proc f(arg: integral) { }
f(true);
```

Note that `integral` means any `int` or `uint` of any width.

`true` can implicitly convert to any `int` width. So, should `f` be instantiated and `true` implicitly converted? Or, should the call fail to resolve?

Today, it does not resolve, but if we write instead `int(?w)`, it does (due to an oddity of how the compiler is implemented):
``` chapel
proc f(arg: int(?w)) { }
f(true);
```

We would like to make these two cases more consistent and remove the implementation oddity, and so need to resolve the question.

### What's the history of the feature, if it already exists?

The production compiler currently stamps out `int(?w)` functions for each width early in compilation. I suspect (but do not know for sure) that it was done this way to bootstrap the compiler development effort, so things like `+` on integers could work even before generics were implemented. But, this stamping-out means that implicit conversions are currently allowed in those forms.

We also currently support instantiation + implicit conversion for class types. For example:

``` chapel
class Parent { type t; }
class Child : Parent { }
proc process(arg: borrowed Parent) { }
proc main() {
  var c = new Child(int);
  process(c); // OK: `process` is instantiated with `Parent(int)`
              // and `c` is implicitly converted into `borrowed Parent(int)`
}
```

This support also extends to class memory management (e.g. it could be instantiating the memory management & implicitly converting from Child to Parent).

### What's the precedent in other languages, if they support it?

I investigated this aspect of C++, Rust, Swift, and Julia. None of them seem to have the capability to do instantiation and implicit conversion for numeric types.

<details>

#### C++

Looks to me like C++ does not allow it.

This program does not compile (indicating that `f(T, T)` cannot be instantiated with `float` and then implicitly convert the 2nd actual argument that is an `int`).

``` c++
template<typename T>
void f(T, T) { }

int main() {
  f(1.0, 1); // error: can't figure out what T should be
  return 0;
}
```

Here is an example C++20 program using concepts which also does not compile:
``` c++
#include <type_traits>

template<class T>
concept FloatingPoint = std::is_floating_point<T>::value;

template<FloatingPoint T>
void fooConcept(T arg) {
}

void fooConcrete (double arg) {
}

int main() {
  fooConcrete(1);.
  fooConcept(1); // error: 'int' does not satisfy FloatingPoint
}
```


#### Rust

Rust doesn't have implicit conversions for numeric types.

``` rust
fn test(x: i32) { }
fn main() {
  let i = 0_i8;
  test(i); // type error
}
```

#### Swift

Does not seem to be allowed in Swift

``` swift
func foo(arg: any BinaryInteger) { }

var s = true;
foo(arg: s);
```

#### Julia

Does not seem to be allowed in Julia

``` julia
function foo(AbstractFloat x)
end
foo(1) // error: no method matching foo(::Int64)
foo(1.0) // OK
```

</details>

### Known mentions of this issue / related features

A bit of a side-issue since it's about user-defined implicit conversions rather than the conversions among the built-in numeric types. It's hard to see how to make a generic Matrix type that can implicitly convert to a generic Array type when passed to a function expecting a generic Array type:

https://github.com/chapel-lang/chapel/issues/5054#issuecomment-282748366

An example from the CHIP 2 proposal seems to imply instantiation+implicit conversion for a numeric type:
https://github.com/chapel-lang/chapel/issues/20011#issuecomment-1613433700

### Proposals

#### Option 1

Continue with the status quo; implicit subtyping conversions are allowed (to support passing a subclass to a superclass) but instantiation + implicit conversion is not allowed for numeric types.

#### Option 2

Allow implicit conversions for numeric types when instantiating built-in type classes such as `integral` or a future generic implementation of `int(?w)`.

**Comment by mppf** [comment]
Based upon discussion in a design subteam (consisting of @vasslitvinov @bradcray and myself), we have decided upon Option 1: to make formals like `a: int(?w)` no longer allow for implicit conversion and instead to be more similar to formals like `a: integral` which do not currently allow for implicit conversion.

The rationale for this is primarily that it is a more conservative design. We are concerned that the alternative has more potential for surprise and confusion. In particular, if we choose Option 1, someone confused about the behavior — say they think the language uses Option 2 — will probably get a compilation error about a failure to resolve a call, which is relatively straightforward to fix. In contrast, if we were to choose Option 2, and somebody expected Option 1, they would get implicit conversions in ways they did not expect, possibly causing overload selection to have different and confusing behavior. In other words, unexpected implicit conversions are a harder to understand and debug failure mode.

One aspect of Option 1 is that it treats implicit conversions that support subtyping for classes differently from the implicit conversion for numeric types. However we are comfortable with that since 1) it's clear enough where to draw the line using the idea of subtype that we already have (which comes up in answer to the question of - what can be passed to `const ref arg: T`?) and 2) classes are special and it's OK for the language to have some special rules for them.

It was interesting to us in the language survey that Rust, Swift, and Julia don't have implicit conversions for numeric variables at all. That gives us more desire to investigate if Chapel could operate without implicit numeric conversions (perhaps allowing implicit param conversions only). Option 1 is slightly closer to this possibility because it will cause code working with formals with types along the lines of `real(?w)` to stop relying on implicit conversions.

We did have these concerns about Option 1, but they were not enough to convince us to go with Option 2:
 * it introduces a discontinuity. If somebody has `proc f(arg: int)`, then they can call it with `f(true)`. But, say they wish to generalize it to accept more integer types. If they write `proc f(arg: int(?w))`, now `f(true)` will no longer resolve.
 * ideally, classes and non-classes would behave the same in terms of implicit conversions
